Source/WebKit/WPEPlatform/wpe/wayland/WPEViewWayland.cpp

```diff
@@ -399,6 +405,55 @@ static const struct wl_buffer_listener cursorBufferListener = {
     }
 };
 
+static const struct zwp_relative_pointer_v1_listener relativePointerListener = {
+    // relative_motion
+    [](void* data, struct zwp_relative_pointer_v1*, uint32_t, uint32_t, wl_fixed_t, wl_fixed_t, wl_fixed_t deltaX, wl_fixed_t deltaY) {
+        auto* view = reinterpret_cast<WPEView*>(data);
+        auto pointerCoords = WPE_VIEW_WAYLAND(view)->priv->pointerCoords;
+        double x = pointerCoords.first;
+        double y = pointerCoords.second;
+        double dX = wl_fixed_to_double(deltaX);
+        double dY = wl_fixed_to_double(deltaY);
+
+        auto* event = wpe_event_pointer_move_new(WPE_EVENT_POINTER_MOVE, view, WPE_INPUT_SOURCE_MOUSE, 0, static_cast<WPEModifiers>(0), x, y, dX, dY);
+        wpe_view_event(view, event);
+        wpe_event_unref(event);
+    }
+};
+
+static gboolean wpeViewWaylandLockPointer(WPEView* view)
+{
+    auto* priv = WPE_VIEW_WAYLAND(view)->priv;
+    if (priv->relativePointer || priv->lockedPointer)
+        return FALSE;
+
+    auto* display = WPE_DISPLAY_WAYLAND(wpe_view_get_display(view));
+    auto* pointerConstraints = wpeDisplayWaylandGetPointerConstraints(display);
+    auto* relativePointerManager = wpeDisplayWaylandGetRelativePointerManager(display);
+    if (!pointerConstraints || !relativePointerManager)
+        return FALSE;
+
+    WPE::WaylandSeat* seat = wpeDisplayWaylandGetSeat(display);
+    priv->pointerCoords = seat->pointerCoords();
+    struct wl_pointer* wlPointer = wl_seat_get_pointer(seat->seat());
+
+    priv->relativePointer = zwp_relative_pointer_manager_v1_get_relative_pointer(relativePointerManager, wlPointer);
+    zwp_relative_pointer_v1_add_listener(priv->relativePointer, &relativePointerListener, view);
+    auto* wlSurface = wpe_view_wayland_get_wl_surface(WPE_VIEW_WAYLAND(view));
+    priv->lockedPointer = zwp_pointer_constraints_v1_lock_pointer(pointerConstraints, wlSurface, wlPointer, nullptr, ZWP_POINTER_CONSTRAINTS_V1_LIFETIME_PERSISTENT);
+
+    return TRUE;
+}
+
+static gboolean wpeViewWaylandUnlockPointer(WPEView* view)
+{
+    auto* priv = WPE_VIEW_WAYLAND(view)->priv;
+    g_clear_pointer(&priv->lockedPointer, zwp_locked_pointer_v1_destroy);
+    g_clear_pointer(&priv->relativePointer, zwp_relative_pointer_v1_destroy);
```

Comment:
These should be done in dispose too, in case the view is destroyed while pointer is locked.